### PR TITLE
qt: multiple fixes

### PR DESCRIPTION
**Pull request by thiagokokada**
### Description

Remove remaining Qt 4 support from `qt` module since `nixpkgs` does not support Qt 4 anymore.

Export `QT_PLUGIN_PATH` and `QML2_IMPORT_PATH`, necessary for some plugins to work (e.g. `qt6ct`).

Remove top-level `with lib`.

Add missing style mappings.

Add `lxqt` support.

### Checklist

- [x] Change is backwards compatible.
  + Removal of Qt 4 technically is not backwards compatible, but I don't think nobody cares.

- [x] Code formatted with `./format`.

- [ ] Code tested through `nix-shell --pure tests -A run.all` or `nix develop --ignore-environment .#all` using Flakes.

- [x] Test cases updated/added. See [example](https://github.com/nix-community/home-manager/commit/f3fbb50b68df20da47f9b0def5607857fcc0d021#diff-b61a6d542f9036550ba9c401c80f00ef).

- [x] Commit messages are formatted like

    ```
    {component}: {description}

    {long description}
    ```

    See [CONTRIBUTING](https://github.com/nix-community/home-manager/blob/master/docs/contributing.adoc#sec-commit-style) for more information and [recent commit messages](https://github.com/nix-community/home-manager/commits/master) for examples.

- If this PR adds a new module

  - [ ] Added myself as module maintainer. See [example](https://github.com/nix-community/home-manager/blob/068ff76a10e95820f886ac46957edcff4e44621d/modules/programs/lesspipe.nix#L6).

### CODE REVIEW

File: modules/misc/qt.nix
Comment (thiagokokada) [inline modules/misc/qt.nix:148]: Tested in my config:

```
$ echo $QT_PLUGIN_PATH
/nix/store/hdvlnzaxdm06xkipg62mj30aa3fzi5fa-fcitx5-with-addons-5.1.1/lib/qt-6/plugins:/etc/profiles/per-user/thiagoko/lib/qt-5.15.10/plugins:/etc/profiles/per-user/thiagoko/lib/qt-6/plugins
$ echo $QML2_IMPORT_PATH
/etc/profiles/per-user/thiagoko/lib/qt-5.15.10/qml:/etc/profiles/per-user/thiagoko/lib/qt-6/qml
```

Seems to be working fine, including properly keeping the current plugins in PATH.

**Review by thiagokokada — COMMENTED** [review]
CC @ncfavier.

**Comment by thiagokokada** [comment]
I added a workaround since https://github.com/nix-community/home-manager/pull/4468 we also set the `QT_PLUGIN_PATH` when `i18n.inputMethod.enabled = 'fcitx5'.

File: modules/misc/qt.nix
Comment (thiagokokada) [inline modules/misc/qt.nix:198]: I know that using `home.sessionVariables` here is not ideal, but this is the only solution I found for now.

We probably need something similar to `environment.profileRelativeEnvVars` or `environment.profileRelativeSessionVariables` in NixOS to fix this issue once for all.

Comment (thiagokokada) [reply modules/misc/qt.nix:198]: This is at least better than the previous solution I did: https://github.com/nix-community/home-manager/pull/4579/commits/3452e14ec729e0582a5fb120c14ed5c6d7fb64d3

Comment (thiagokokada) [reply modules/misc/qt.nix:198]: > We probably need something similar to environment.profileRelativeEnvVars or environment.profileRelativeSessionVariables in NixOS to fix this issue once for all.

Or even simpler, as long as we have something that allows a similar semantic as `home.sessionPath`, but generic (e.g.: `home.searchPathVariables.QT_PLUGIN_PATH = [ ... ]`), this would already work.

Comment (thiagokokada) [reply modules/misc/qt.nix:198]: I decided to fix this issue in a separate PR: https://github.com/nix-community/home-manager/pull/4582.

I think this PR can still be merged before that PR above. In any case once the other PR is accepted I will change this PR to re-implement this in terms of `home.sessionSearchVariables`.

**Comment by 9p4** [comment]
Does this supersede https://github.com/nix-community/home-manager/pull/4306

**Comment by thiagokokada** [comment]
> Does this supersede #4306

I would say so.

While technically the approach from #4306 is more flexible, I am not sure it makes sense to have those 2 environment variables configurable, except to workaround the composition issue that should be fixed in #4582.

Also this PR fixes a bunch of other stuff (like removal of Qt 4) and adds some other stuff (the LXQt), so I would say this PR pretty much supersede #4306 in many ways.

CC @rycee.

File: modules/misc/qt.nix
Comment (9ary) [inline modules/misc/qt.nix:34]: A bit iffy on this, considering upstream is discontinued (https://github.com/FedoraQt/adwaita-qt#adwaita-qt-project-is-unmaintained-a-no-longer-actively-developed).

I also added it in #4312, but that was a few days before the notice was put up.

File: modules/misc/qt.nix
Comment (9ary) [inline modules/misc/qt.nix:7]: If you're going to drop `with`, I think you should drop all instances of it.

File: modules/misc/qt.nix
Comment (9ary) [inline modules/misc/qt.nix:205]: Nice refactor, nicer than my own attempt at adding the gtk3 platform theme.

**Review by 9ary — COMMENTED** [review]
LGTM overall.

File: modules/misc/qt.nix
Comment (thiagokokada) [reply modules/misc/qt.nix:7]: In general `with` is discouraged at the top level but a scoped one like this usage is generally considered fine.

Also I think this is a by maintainer decision and I am the maintainer of this module, so unless there is a good reason I prefer to keep it as is.

File: modules/misc/qt.nix
Comment (thiagokokada) [reply modules/misc/qt.nix:34]: I don't see a good reason to remove this unless we remove the package from nixpkgs itself.

I also recently added this same option in the NixOS module (see https://github.com/NixOS/nixpkgs/pull/262051), and it would be good to keep for feature parity.

If we eventually remove this from nixpkgs, I think we can remove it from here.

Comment (9ary) [reply modules/misc/qt.nix:34]: Agreed.

**Comment by thiagokokada** [comment]
@9ary I backported your fixes from #4312, can you re-review?

File: modules/misc/qt.nix
Comment (9ary) [reply modules/misc/qt.nix:7]: Sorry, that was just a suggestion.

>a scoped one like this usage is generally considered fine

It's more that it gets used like this *in spite* of the pitfalls, because it's convenient. Either way, I'm not here to bikeshed, so it's your call.

File: modules/misc/qt.nix
Comment (9ary) [inline modules/misc/qt.nix:214]: Not familiar with the xsession stuff (I use wayland). Do we want `QT_QPA_PLATFORMTHEME` here if we're not setting it?

**Review by 9ary — COMMENTED** [review]
Looks fine.

For the gtk3 platform theme, like I said in #4312, I'm not sure if there are any extra dependencies that aren't automatically pulled in.
Setting the variable is all I need on my setup, but I'm not sure how to test this in isolation.

File: modules/misc/qt.nix
Comment (thiagokokada) [reply modules/misc/qt.nix:214]: I think there is no harm setting it if it is not set, but not sure either.

Comment (9ary) [reply modules/misc/qt.nix:214]: Oddly enough that option doesn't show up in the docs, but as far as I can tell it just passes the variable name to `systemctl --user import-environment` in ~/.xsession.

Looking at [the man page](https://www.freedesktop.org/software/systemd/man/latest/systemctl.html#%0A%20%20%20%20%20%20%20%20%20%20%20%20import-environment%0A%20%20%20%20%20%20%20%20%20%20%20%20VARIABLE%E2%80%A6%0A%20%20%20%20%20%20%20%20%20%20):
>Import all, one or more environment variables set on the client into the systemd manager environment block. If a list of environment variable names is passed, client-side values are then imported into the manager's environment block. If any names are not valid environment variable names or have invalid values according to the rules described above, an error is raised.

I'm not entirely sure how to interpret this, what constitutes an invalid variable name?

Empirically:
```
$ systemctl --user import-environment arstarstarstarst
Environment variable $arstarstarstarst not set, ignoring.
```
It just prints a warning and exits with status code 0, so I think it doesn't strictly matter, but I would still gate this to suppress the warning.

Comment (thiagokokada) [reply modules/misc/qt.nix]: Done.

**Review by 9ary — APPROVED** [review]

**Comment by Aleksanaa** [comment]
lgtm

**Comment by thiagokokada** [comment]
CC @ncfavier @rycee, can we have this merged?

**Comment by rycee** [comment]
Thanks! I've merged this now 🙂

**Comment by acristoffers** [comment]
Having `qt.enable = true;` in my config is making it fail with 

```
error:
       … while evaluating a branch condition

         at /nix/store/rj6pqzx66pliiwdfk2wjnnlk40rm3mpd-source/lib/lists.nix:57:9:

           56|       fold' = n:
           57|         if n == len
             |         ^
           58|         then nul

       … while calling the 'length' builtin

         at /nix/store/rj6pqzx66pliiwdfk2wjnnlk40rm3mpd-source/lib/lists.nix:55:13:

           54|     let
           55|       len = length list;
             |             ^
           56|       fold' = n:

       (stack trace truncated; use '--show-trace' to show the full trace)

       error: value is null while a string was expected
```

using a `home.nix` as simple as

```nix
{ config, inputs, pkgs, username, isLinux, ... }:

let
  homeDirectory = if isLinux then "/home/" + username else "/Users/alan";
in
{
  programs.home-manager.enable = true;
  home.stateVersion = "22.11";
  home.username = username;
  home.homeDirectory = homeDirectory;

  qt.enable = true;
}
```

The enterity of the original file and the flake are in my dotfiles repo: https://github.com/acristoffers/dotfiles

**Comment by acristoffers** [comment]
When using `--show-trace` the last error is

```
       … while calling anonymous lambda

         at /nix/store/f4gw9kx4xyjs1i1jxnlr077v7l6lavlh-source/modules/misc/qt.nix:161:35:

          160|     # Necessary because home.sessionVariables doesn't support mkIf
          161|     envVars = lib.filterAttrs (n: v: v != null) {
             |                                   ^
          162|       QT_QPA_PLATFORMTHEME =

       error: value is null while a string was expected
```

**Comment by thiagokokada** [comment]
> When using `--show-trace` the last error is
> 
> ```
>        … while calling anonymous lambda
> 
>          at /nix/store/f4gw9kx4xyjs1i1jxnlr077v7l6lavlh-source/modules/misc/qt.nix:161:35:
> 
>           160|     # Necessary because home.sessionVariables doesn't support mkIf
>           161|     envVars = lib.filterAttrs (n: v: v != null) {
>              |                                   ^
>           162|       QT_QPA_PLATFORMTHEME =
> 
>        error: value is null while a string was expected
> ```

Will take a look later.

**Pull request merged**
